src/config.c

```diff
@@ -767,8 +865,15 @@ static int cgroup_config_create_groups(void)
 
 		error = cgroup_create_cgroup(cgroup, 0);
```

Comment:
Would it make more sense to try and convert the cgroup prior to calling `cgroup_create_cgroup()`.  It feels weird to try and write it, watch it fail, then convert it and try again.

What if we added a flag to the cgconfig file like this:
```
group foo {
        version = 1;
        cpu {
                cpu.shares = 512;
                cpu.cfs_quota_us = 5000;
        }
        cpuset {
                cpuset.cpus = 0-1;
        }
}
```

Then the convert logic would look something like:
```
mode = get_mode();
/// do some trickery to map mode to cgroup version
/// if mode == hybrid, probably skip the conversion because hybrid is weird and hard to work with

if system_cg_version != config_cg_version:
    cgroup_convert_cgroup();

cgroup_create_cgroup()
```

Reply:
Actually we could probably handle the hybrid case as well if we really wanted to.  We might have to split up the cgroup if it has different target versions (say cpu controller is v1 and memory controller is v2).  Or another idea would be to expose `convert_controller()` outside of abstraction-common.c

Reply:
> Would it make more sense to try and convert the cgroup prior to calling cgroup_create_cgroup(). It feels weird to try and write it, watch it fail, then convert it and try again.

It's an interesting idea to use hints.  I guess the commit explanation is horrible in doing its job of explaining the approach, please allow me to explain the idea again and let me know, your thoughts on the approach, consider following the `cgconfig.conf`
```
group foo {
   cpuset {
           cpuset.cpus = 0;
   }
   cpu {
           cpu.shares = 512;
   }
}
```
parsing of the above file by the `cgconfigparser` will fail on cgroup v2 setup, since `cpu.shares` is cgroup v1 `cpu` controller setting.  It is easy to find the incorrect settings by glancing over the `cgconfig.conf` file but it gets difficult to guess the controller with incorrect settings during creation of cgroup `foo` with `cpu, cpuset` controllers from `cgconfigparser` code path.  The crux of the approach is: 
(for the sake of readability, `src_cgrp`  refers to the cgroup created by the `configparser` and `dst_cgrp`  refers to the cgroup created by the conversion code.)
1. Create `dst_cgrp` and  copy the permissions from the `src_cgrp`
2. Loop over `src_cgrp->controller[index]` and attach the `ith` controller (attach only one controller)
3.  Try creating the `dst_cgrp` with only `ith` controller from `src_cgrp`, and if it succeeds, move on to the next controller.  With the above example, `src_dst->index == 0` is `cpuset` controller, and `dst_cgrp` is created with only `cpuset` controller succeeds.
4. If step 3 succeeds,  the `dst_cgrp->controller`  index is reset to zero, (resetting `dst_cgrp` to empty cgroup with no controllers) and the single allocated controller is freed.  In our example, it's the  `cpuset` controller, which gets free'd, and repeat the same for the `i+1` controller of `src_cgrp` by starting from step 2.
5. When step 3  fails, conversion is attempted for the `dst_cgrp`,  with the controllers `convert_version` derived from simple boolean logic, i.e, if the current controller version is cgroup v1, then cgroup v2 is the `convert_version` and vice versa.  The logic is independent of checking the current cgroup setup mode to derive the `convert_version`.
6. If step 5 succeeds, repeat the steps from step 2 for the `i+1` controller or execute the error path on failing to convert the `dst_cgrp` controller setting to `convert_version`.

previously any one incorrect setting of a controller in the `cgconfig.conf`, will fail to create the cgroup.  But in the above patch,  the cgroup might exist with other controllers other than the first controller with incorrect settings, depending upon the order of controllers as parsed by the `cgconfigparser`.  If the above `cgconfig.conf` is modified to:
```
group foo {
   cpuset {
           cpuset.cpus = 0;
   }
   cpu {
           cpu.unknown = 512;
   }
}
```
in the above example, cgroup `foo` will exist with `cpuset` controller enabled and will execute the error path on the failed conversion of `cpu.unknown` setting of the `cpu` controller.  Such failure can be handled by either falling back to the original behavior of not creating the cgroup when an incorrect setting is encountered or skipping the faulty controller and creating the cgroup with the good controllers.  In the above patch, neither of the approaches is used and it needs to be fixed.

Reply:
Actually returning the nonzero will call `cgroup_config_destroy_groups()` in the error path, cleaning up the cgroups created by the `cgconfig.conf` on the first occurrence of error/wrong/unconvertable cgroup controller settings.